### PR TITLE
Add preview to datasets as specified in the Kedro catalog under metadata

### PR DESCRIPTION
**Pull request by rashidakanchwala**
## Description

This ticket is part 2 of the preview dataset feature development. In the first part, we enabled preview functionality for CSV and Excel datasets in the Kedro Framework. In Kedro-viz, we showed previews for all CSV and Excel datasets when the user turned on the preview option under the settings panel.

In this ticket, we are removing the preview option from the settings panel. Now, if the user wants to preview a dataset, they need to specify it via the catalog.yml file.

```yaml
type: pandas.CSVDataSet
filepath: /filepath/to/dataset
metadata:
  kedro-viz:
    layer: raw
    preview_args:
          nrows: 3
```

In the above the user wants to see the first 3 rows of the dataset. Previously we always showed the first 40 rows, but now we show the number of rows specified by the user in the catalog.yml file.

## QA notes

The tests and linter is currently failing but the changes to this are already part of the first PR related to layers so it should be fixed once that's merged. So we can ignore that in this PR. 

## Checklist

- [ ] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added new entries to the `RELEASE.md` file
- [ ] Added tests to cover my changes

<a href="https://gitpod.io/#https://github.com/kedro-org/kedro-viz/pull/1374"></a>

### CODE REVIEW

File: src/components/metadata-modal/metadata-modal.js [hunk 1]
```diff
@@ -46,7 +46,7 @@ const MetadataModal = ({ metadata, onToggle, visible }) => {
         </div>
         {hasPreview && (
           <div className="pipeline-metadata-modal__preview-text">
-            Previewing first 40 rows only
+            Previewing first {metadata.preview.data.length} rows only
```
Comment (Huongg) [inline src/components/metadata-modal/metadata-modal.js:49]: I think this copy line might not make sense anymore. We use it originally because we were only going to show the first 40 rows. But now since users can define it themself, do we need it anymore?

File: demo-project/conf/base/catalog_01_raw.yml [hunk 2]
```diff
@@ -1,14 +1,41 @@
...
-  layer: raw
+  metadata:
+    kedro-viz:
+      layer: raw
+      preview: 15 
+
+
+
+
+# companies:
```
Comment (Huongg) [inline demo-project/conf/base/catalog_01_raw.yml:28]: are we not going to remove these completely? if not then it might be helpful to have a comment to say why we just commented them out here

**Comment by antonymilne** [comment]
My suggested syntax for this is:
```
type: pandas.CSVDataSet
filepath: /filepath/to/dataset
metadata:
  kedro-viz:
    layer: raw
    preview_args:
      nrows: 5
```

This would match `load_args` and `save_args` and mean that future preview functions can take an arbitrary set of arguments. Otherwise you're hardcoding it here to only accept `nrows` as a parameter.

You'd call the preview function as `dataset._preview(**preview_args)`.

There's a question about what the default behaviour should be though. If I don't specify `preview_args`, should the preview show at all?
* if yes, then should a user be able to turn it off if they like and how would they do this?
* if no, then you always need to specify `nrows` explicitly, which is annoying.

**Review by Huongg — COMMENTED** [review]
hey @rashidakanchwala all looks good to me, just left 2 small comments but nothing seems to be an issue to me.

**Comment by merelcht** [comment]
I agree with @antonymilne 's comment above. `preview_args` would be much more flexible and allow for other things than just number of rows to be shown in the preview. 

> There's a question about what the default behaviour should be though. If I don't specify preview_args, should the preview show at all?

IMO if the user doesn't specify `preview_args` on the dataset it shouldn't preview at all. It seems confusing if it would show without adding anything in the catalog. Also, if we add more options and not just number of rows, what would you show by default?

**Comment by rashidakanchwala** [comment]
> My suggested syntax for this is:
> 
> ```
> type: pandas.CSVDataSet
> filepath: /filepath/to/dataset
> metadata:
>   kedro-viz:
>     layer: raw
>     preview_args:
>       nrows: 5
> ```
> 
> This would match `load_args` and `save_args` and mean that future preview functions can take an arbitrary set of arguments. Otherwise you're hardcoding it here to only accept `nrows` as a parameter.
> 
> You'd call the preview function as `dataset._preview(**preview_args)`.
> 
> There's a question about what the default behaviour should be though. If I don't specify `preview_args`, should the preview show at all?
> 
> * if yes, then should a user be able to turn it off if they like and how would they do this?
> * if no, then you always need to specify `nrows` explicitly, which is annoying.

The preview shouldn't show at all. Basically, if you want to show preview then you need to have the preview_args in your paml and nrows would follow. If people don't specify nrows but want to still preview what would it look like. I am a bit confused. 

> ```
> type: pandas.CSVDataSet
> filepath: /filepath/to/dataset
> metadata:
>   kedro-viz:
>     layer: raw
>     preview_args: ?
> ```

Currently, we don't have any other options. But I guess if the user doesn't add nrows, it would default to 40, which is what is specified already in kedro-datasets. I will still test it

File: package/kedro_viz/models/flowchart.py [hunk 3]
```diff
@@ -488,7 +488,12 @@ def is_tracking_node(self):
 
     def is_preview_node(self):
         """Checks if the current node has a preview"""
-        return hasattr(self.kedro_obj, "_preview")
+        metadata = getattr(self.kedro_obj, "metadata", {}) or {}
+        return bool(metadata.get("kedro-viz", {}).get("preview"))
```
Comment (noklam) [inline package/kedro_viz/models/flowchart.py:492]: It feels like it shouldn't be that complicated. 
Many questions here:
* Does `preview: 0` count as `True` or `False`? It will be `False` now, then the `get("preview", 0) path will never be executed.

The same pattern accessing the nested dict appears in multiple places, maybe refactor it with a small util function

**Comment by antonymilne** [comment]
@merelcht 
> > There's a question about what the default behaviour should be though. If I don't specify preview_args, should the preview show at all?
> 
> IMO if the user doesn't specify `preview_args` on the dataset it shouldn't preview at all. It seems confusing if it would show without adding anything in the catalog. Also, if we add more options and not just number of rows, what would you show by default?

I think the problem with this approach is that the user always has to specify `nrows`. Currently this has a default value in kedro-datasets:
```
_preview(self, nrows: int = 40)
```
... but as it stands there's no way to actually use that default value, since as soon as you write `preview_args` then you need to specify `nrows`. In theory it might be nice for a user to just switch `preview` on but without needing to supply any more information.

For the case of pandas datasets it's not too much effort for the user, but in future maybe we'll have `preview` methods for lots of different datasets. In that case it would be nice for a user to just be able to say `preview: True` without needing to specify all the `preview_args` which might well be different for different dataset tyes.

It's not a big thing so I'm ok with this if it's what we want to do, but I'd probably make `nrows` compulsory in kedro-datasets in this case.

The alternative solution is maybe we have two keys: `preview` and `preview_args`, like this:
```
preview: true
preview_args:
  nrows: ...
```
where `preview_args` is optional and only for "advanced" users who want to fine-tune the preview shown.

Or another option is that `preview` is enabled by default and they need to specify `preview: false` to disable it.

**Comment by merelcht** [comment]
@antonymilne We discussed this in stand up last week and agreed that the user needs to be specific about what needs to be previewed. So they will have to specify `preview_args` or nothing will show. There just won't be a default at all. We can always iterate on that solution once we get user feedback on their experience, but at the moment we don't even know yet what potential other preview functions we or other people will add.

File: src/components/metadata-modal/metadata-modal.js (hunk 1 above)
Comment (merelcht) [reply src/components/metadata-modal/metadata-modal.js:49]: I think it's still useful to have the message but maybe rephrase it to: "Previewing first ... rows"

Comment (tynandebold) [reply src/components/metadata-modal/metadata-modal.js:49]: Agree with Merel here. Let's still give them the number.

File: package/kedro_viz/models/flowchart.py [hunk 4]
```diff
@@ -486,14 +486,28 @@ def is_tracking_node(self):
...
+    def get_viz_metadata(self):
+        """Gets the kedro-viz metadata config specified in the catalog"""
+        metadata = getattr(self.kedro_obj, "metadata", None)
+        if not metadata:
+            return None
+        try:
+            viz_metadata = metadata["kedro-viz"]
+        except (AttributeError, KeyError):
+            return None
+        return viz_metadata
```
Comment (noklam) [inline package/kedro_viz/models/flowchart.py:498]: Can this be replaced with the `getattr` method since it's not a nested key?

Comment (noklam) [reply package/kedro_viz/models/flowchart.py:498]: additional comment: Would it be better to just set this as a property `viz_metadata` instead of a get method?

Comment (rashidakanchwala) [reply package/kedro_viz/models/flowchart.py:498]: i did try this but I got some weird errors, so changed it to get_ . but i agree, it's better as a property. I will ping u again, if u have a couple of min to help me with the errors that would be appreciated. let me try doing it again.

File: demo-project/conf/base/catalog_01_raw.yml [hunk 5]
```diff
@@ -1,14 +1,29 @@
 companies:
   type: pandas.CSVDataSet
   filepath: ${base_location}/01_raw/companies.csv
-  layer: raw
+  metadata:
+    kedro-viz:
+      layer: raw
+      preview_args:
```
Comment (merelcht) [inline demo-project/conf/base/catalog_01_raw.yml:7]: I think there's a value missing here?

File: demo-project/conf/base/catalog_04_feature.yml [hunk 6]
```diff
@@ -29,6 +29,9 @@
 feature_importance_output: 
   type: pandas.CSVDataSet
   filepath: ${base_location}/04_feature/feature_importance_output.csv
-  layer: feature 
+  metadata:
+    kedro-viz:
+      layer: feature 
+      preview: 2
```
Comment (merelcht) [inline demo-project/conf/base/catalog_04_feature.yml:35]: I thought we would just do the `preview_args` and not the `preview`?

File: package/kedro_viz/models/flowchart.py [hunk 7]
```diff
@@ -458,6 +460,12 @@ def __post_init__(self):
         self.modular_pipelines = self._expand_namespaces(
             self._get_namespace(self.full_name)
         )
+        metadata = getattr(self.kedro_obj, "metadata", None)
+        if metadata:
+            try:
+                self.viz_metadata = metadata["kedro-viz"]
+            except (AttributeError, KeyError):
+                pass
```
Comment (merelcht) [inline package/kedro_viz/models/flowchart.py:468]: Would it be helpful here to emit an error message for debugging purposes?

**Review by merelcht — COMMENTED** [review]
@rashidakanchwala I'll message you on Slack as well because I'm a bit confused now about the implementation: are we using `preview_args` or `preview` with a numeric value? It seems mixed in the changes you've made..

File: demo-project/conf/base/catalog_04_feature.yml (hunk 6 above)
Comment (rashidakanchwala) [reply demo-project/conf/base/catalog_04_feature.yml:35]: this is just old code. i removed it. thanks for flagging

File: package/kedro_viz/models/flowchart.py [hunk 8]
```diff
@@ -488,7 +497,15 @@ def is_tracking_node(self):
...
         """Checks if the current node has a preview"""
-        return hasattr(self.kedro_obj, "_preview")
+        try:
+            is_preview = bool(self.viz_metadata["preview_args"])
+        except (AttributeError, KeyError):
+            return False
+        return is_preview
+
+    def get_preview_args(self):
+        """Gets the number of rows for the preview dataset"""
```
Comment (merelcht) [inline package/kedro_viz/models/flowchart.py:507]:
```suggestion
        """Gets the preview arguments for a dataset"""
```

File: package/tests/test_models/test_flowchart.py [hunk 9]
```diff
@@ -368,6 +369,17 @@ def test_data_node_metadata(self):
         assert data_node_metadata.filepath == "/tmp/dataset.csv"
         assert data_node_metadata.run_command == "kedro run --to-outputs=dataset"
 
+    def test_get_preview_nrows(self):
+        metadata = {"kedro-viz": {"preview": 3}}
```
Comment (merelcht) [inline package/tests/test_models/test_flowchart.py:373]: I'm guessing this test needs to be updated to use `preview_args` right?

**Review by merelcht — COMMENTED** [review]
Almost there!

File: package/tests/test_models/test_flowchart.py (hunk 9 above)
Comment (rashidakanchwala) [reply package/tests/test_models/test_flowchart.py:373]: thanks for this. i missed it. the tests are still failing but that's because of the kedro.extras and it should go once the other PR is merged.

**Review by merelcht — APPROVED** [review]
Nice work! ⭐ Don't forget to add the change to the release notes.

File: package/tests/test_models/test_flowchart.py [hunk 10]
```diff
@@ -368,6 +369,18 @@ def test_data_node_metadata(self):
...
+    def test_get_preview_args(self):
+        metadata = {"kedro-viz": {"preview_args": {"nrows": 3}}}
+        dataset = CSVDataSet(filepath="test.csv", metadata=metadata)
+        data_node = GraphNode.create_data_node(
+            full_name="dataset",
+            tags=set(),
+            layer=None,
+            dataset=dataset,
+        )
+        assert data_node.is_preview_node() is True
```
Comment (antonymilne) [inline package/tests/test_models/test_flowchart.py:381]:
```suggestion
        assert data_node.is_preview_node()
```

**Review by antonymilne — APPROVED** [review]
A few small suggestions but generally this looks great ⭐

File: package/kedro_viz/models/flowchart.py [hunk 11]
```diff
@@ -488,7 +496,15 @@ def is_tracking_node(self):
 
     def is_preview_node(self):
         """Checks if the current node has a preview"""
-        return hasattr(self.kedro_obj, "_preview")
+        try:
+            is_preview = bool(self.viz_metadata["preview_args"])
+        except (AttributeError, KeyError):
+            return False
+        return is_preview
```
[the remark this answers is not in the document]
Comment (antonymilne) [reply package/kedro_viz/models/flowchart.py:503]: btw I see this is probably the opposite of what @noklam was suggesting, so feel free to ignore and just do whatever you think is best 😁

[the remark this answers is not in the document]
Comment (rashidakanchwala) [reply package/kedro_viz/models/flowchart.py:503]: I think viz_metadata might be eventually used for more than preview. for e.g. looking at this (https://github.com/kedro-org/kedro-viz/issues/662) ticket. if we were to also show table statistics we could probably have 

metadata :
   kedro-viz: 
       preview_args : 
             nrows=4 
      profiler_args:
              show = true.. 
              
something like it
